### PR TITLE
Improve DOMTokenList performance

### PR DESCRIPTION
**Pull request by flagello**
I was hoping to borrow some classes here and there — while I was there I thought why not.

### CODE REVIEW

File: src/DOMTokenList.js [hunk 1]
```diff
@@ -27,7 +27,22 @@ module.exports = class DOMTokenList extends Array {
...
+    const adding = tokens.length;
+    if (adding === 1) {
+      const token = tokens[0];
+      for (let i = 0; i < current; i++) {
+        if (this[i] === token) return;
+      }
+      this.push(token);
+    } else {
+      main: for (let i = 0; i < adding; i++) {
+        for (let j = 0; j < current; j++) {
```
Comment (WebReflection) [inline src/DOMTokenList.js]: O(n^2) ?

File: src/DOMTokenList.js [hunk 2]
```diff
@@ -72,10 +72,13 @@ module.exports = class DOMTokenList extends Array {
   }
 
   replace(token, newToken) {
-    const i = this.indexOf(token);
-    if (i < 0) this.add(newToken);
-    else this[i] = newToken;
-    afterChanges(this);
+    for (let i = 0, n = this.length; i < n; i++) {
```
Comment (WebReflection) [inline src/DOMTokenList.js]: are you avoiding `indexOf` here? If that's the case ... why?

File: src/DOMTokenList.js [hunk 3]
```diff
@@ -72,10 +72,13 @@ module.exports = class DOMTokenList extends Array {
...
+      if (this[i] === token) {
+        this[i] = newToken;
+        return afterChanges(this);
+  }
+    }
+    this.add(newToken);
```
Comment (WebReflection) [inline src/DOMTokenList.js]: or maybe `afterChanges` was triggered twice? curious to know more about this one

File: src/DOMTokenList.js [hunk 4]
```diff
@@ -23,7 +23,10 @@ module.exports = class DOMTokenList extends Array {
   }
 
   contains(token) {
-    return this.includes(token);
+    for (let i = 0, n = this.length; i < n; i++) {
```
Comment (WebReflection) [inline src/DOMTokenList.js]: and here you are avoiding native includes ... I'm not too fund of augmenting complexity for no reason when the platform provides default alternatives

**Comment by flagello** [comment]
I understand entirely @WebReflection. I see your point and I welcome your approach, despite not fully agreeing with it.

_(I'll leave reference replies to the relative comments, but they are by no means a way to dispute your ideas or work. Despite the striking difference: a) I'm not entirely sure how they translate in real-world usage; b) I didn't test them in Node. A contrasting observation might be that the difference could be more significant on mobile.)_

File: src/DOMTokenList.js (hunk 4 above)
Comment (flagello) [inline src/DOMTokenList.js]: https://esbench.com/bench/5942eb3d99634800a0348611

File: src/DOMTokenList.js (hunk 2 above)
Comment (flagello) [inline src/DOMTokenList.js]: https://esbench.com/bench/5942edd099634800a034862a

File: src/DOMTokenList.js (hunk 1 above)
Comment (flagello) [inline src/DOMTokenList.js]: https://esbench.com/bench/5942696599634800a03485a2

File: src/DOMTokenList.js (hunk 4 above)
Comment (WebReflection) [inline src/DOMTokenList.js]: I cannot see anything in that link but regardless includes does more than your code so this change won't land in this project

File: src/DOMTokenList.js (hunk 2 above)
Comment (WebReflection) [inline src/DOMTokenList.js]: I cannot see anything in that link but regardless indexOf is one word to maintain VS few lines so this change won't land in this project

File: src/DOMTokenList.js (hunk 1 above)
Comment (WebReflection) [inline src/DOMTokenList.js]: I cannot see anything in that link but O(n^2) wasn't a question, rather a statement. The algorithm used in here is not even a performance oriented one so I'm afraid this change won't land in this project

File: src/DOMTokenList.js (hunk 4 above)
Comment (flagello) [inline src/DOMTokenList.js]: That's a glaring oversight. I'll be sure to refer to the reference implementation. Thanks for pointing it out!

**Comment by WebReflection** [comment]
> A contrasting observation might be that the difference could be more significant on mobile.

I've been around for too long to smell 10 years ago useless optimizations in here ... the JIT era went way beyond this and modern engines should be able to optimize better than JS itself.

I'm really not coding thinking that I cannot `array.join(...)` and nobody, not even new contributors, should ever feel intimidated by unnecessary bloat for benchmarks that are irrelevant not only in nodejs, but in this whole project scope.

Again, I appreciate your interest to this project but if it's crazy-faster-super-performance you are looking for, you're better off with any of the alternative out there, really: this project is not about crazy-fast performance ... this project is about bringing, with ease to read and contribute, a basic HTML env to any non HTML environment.

On nodeJS I would not use this, unless I've pre-created documents per page ... and ther the serialization of `document.toString()` is already the fastest **and** the only one I care but not at the point I cannot read anymore a bloody `array.join()` :wink:

**Comment by WebReflection** [comment]
P.S. you can fork this project and bring in all the micro-benchmarks you want. Maybe one day all will be merged back together :smiley:

**Pull request closed without merging**